### ISSUE TITLE
[V3 Core] Difficulty reloading a cog with multiple modules. 

### ISSUE THREAD

**Issue by Redjumpman** [Cog-Creators/Red-DiscordBot#1457]
Please be sure to read through other issues as well to make sure what you are suggesting/reporting has not already
been suggested/reported

### Type:

- [ ] Suggestion
- [x] Bug

### Brief description of the problem
Reloading a cog seems to only work with the main module. Auxiliary module changes do not appear to be reflected in the reload. Workaround involves doing a bot shutdown and starting the bot again.

### Expected behavior
The cog and all supporting modules in the folder are "reloaded" to reflect the changes.
### Actual behavior
Only changes made to the main cog are reflected in the main cog. Shutting down the bot and starting it again is the only way to see the changes.
### Steps to reproduce

1. Make a cog with 2 modules
2. Add in an reproducible error to the auxiliary module.
3. Test the error
4. Fix the error
5. Reload the cog

**Comment by Tobotimus** [comment]
This is a pretty complex issue unfortunately, and something lots of projects have trouble with. Here's an example of why.

We have a python package:
```
package
 | - __init__.py
 | - module1.py
 \ - module2.py
```

Say we make some modifications to the code in some of these modules, and want to reload it. Obviously, only one module can be interpreted at a time. So which one do we do first? It begs the question. What the `reload` command does is iterates over each module inside the package, reloading one after the other.

For now let's assume we're reloading `__init__.py` first.

Say inside `__init__.py`, we have this import line:
```py
from .module1 import *
```
So from here, python will go and look for whatever needs to be imported from `module1.py`. However, it's doing this *before we have a chance to reload `module1.py`*, hence whatever is imported will be from the *old module*.

The same thing will happen when we reload `module1.py`, if it depends in any way on `module2.py`.

Reloading a module after a separate module has already imported an object from it obviously will not change the behaviour of that imported object.

### In summary
This issue is dependent on the import order of the modules within your package. Sometimes, reloading will show the changes, sometimes it won't.

To solve this issue, we need to recursively reload modules which are *imported* from within other modules. However this recursion isn't particularly reliable if two modules import each other. You've struck gold on this one, RJM :)

**Comment by tekulvw** [comment]
Should be mostly fixed with #1458, closing for now. If this issue pops up again we'll need another issue with more detail on that specific case.
